x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/latest/LatestDoc.java

```diff
@@ -0,0 +1,280 @@
+        return compositeAgg.getBuckets().stream()
+            .map(bucket -> {
+                transformIndexerStats.incrementNumDocuments(bucket.getDocCount());
+                TopHits topHits = bucket.getAggregations().get(TOP_HITS_AGGREGATION_NAME);
+                assert topHits.getHits().getHits().length == 1;
+                Map<String, Object> document = topHits.getHits().getHits()[0].getSourceAsMap();
+
+                // generator to create unique but deterministic document ids, so we
+                // - do not create duplicates if we re-run after failure
+                // - update documents
+                IDGenerator idGen = new IDGenerator();
+                config.getUniqueKey().forEach(field -> idGen.add(field, bucket.getKey().get(field)));
+
+                document.put(TransformField.DOCUMENT_ID_FIELD, idGen.getID());
+                return document;
+            });
+    }
+
+    @Override
+    public Tuple<Stream<IndexRequest>, Map<String, Object>> processSearchResponse(
+        SearchResponse searchResponse,
+        String destinationIndex,
+        String destinationPipeline,
+        Map<String, String> fieldTypeMap,
+        TransformIndexerStats stats
+    ) {
+        Aggregations aggregations = searchResponse.getAggregations();
+
+        // Treat this as a "we reached the end".
+        // This should only happen when all underlying indices have gone away. Consequently, there is no more data to read.
+        if (aggregations == null) {
+            return null;
+        }
+
+        CompositeAggregation compositeAgg = aggregations.get(COMPOSITE_AGGREGATION_NAME);
+        if (compositeAgg == null || compositeAgg.getBuckets().isEmpty()) {
+            return null;
+        }
+
+        Stream<IndexRequest> indexRequestStream =
+            extractResults(compositeAgg, stats)
+                .map(document -> DocumentConversionUtils.convertDocumentToIndexRequest(document, destinationIndex, destinationPipeline));
+        return Tuple.tuple(indexRequestStream, compositeAgg.afterKey());
+    }
+
+    @Override
+    public void validateQuery(Client client, SourceConfig sourceConfig, ActionListener<Boolean> listener) {
+        SearchRequest searchRequest = buildSearchRequest(sourceConfig, null, TEST_QUERY_PAGE_SIZE);
+
+        client.execute(SearchAction.INSTANCE, searchRequest, ActionListener.wrap(response -> {
+            if (response == null) {
+                listener.onFailure(
+                    new ElasticsearchStatusException("Unexpected null response from test query", RestStatus.SERVICE_UNAVAILABLE)
+                );
+                return;
+            }
+            if (response.status() != RestStatus.OK) {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "Unexpected status from response of test query: " + response.status(),
+                        response.status()
```

Comment:
```suggestion
                        "Unexpected status from response of test query: {}",
                        response.status(),
                        response.status()
```
Doesn't this work?

Reply:
I've copied this line from `Pivot`. Your suggestion is perfectly valid.
I've just changed it both here and in `Pivot`.